### ISSUE TITLE
Versioning / Branching Strategy

### ISSUE THREAD

**Issue by vitaly-livshits** [MicrosoftLearning/AZ-203-DevelopingSolutionsforMicrosoftAzure#63]
Challenge we've just run into this week: lab instructions are cached by the Authorized Lab Hosters (ALHs), while many of the labs direct students to clone the most recent version of the master branch. 

This results in a discrepancy between the version of instructions vs. version of files pulled from GitHub. Thus the labs get broken from student perspective. This week it was Lab 3, however, this could impact any lab.

Possible solution: modify the lab instructions so they clone a specific version instead of the latest. For example, use tags, dates or another other versioning mechanism. Another approach would be to create branches or tags which separate major releases with breaking changes from minor updates without breaking changes.

There may be better options. Feedback welcome.

**Comment by seesharprun** [comment]
Which ALH are you using?

**Comment by vitaly-livshits** [comment]
Learn on Demand Systems

**Comment by tiagocostapt** [comment]
it would be so much easier if ALH's just connected their systems with the public repo... @vitaly-livshits I have the same issue every time a delivered a class. I just redirect my students to the GitHub repo. After all this is a developer course, they should be familiar with GitHub, if not, this is an awesome opportunity. :)

**Comment by seesharprun** [comment]
For LODS, they require that MCTs file a support ticket here: https://www.learnondemandsystems.com/customer-support/ to get the course updated.

Ideally, the ALH partners would download the AllFiles content and put it on the labs and we could remove the steps at the beginning of the lab to clone the GitHub repository. I would love to see those steps removed from the beginning of every lab.

Unfortunately, we learned (in 20532 and 20535) that no-one ever copied the AllFiles content leading to multiple bug reports. This was a compromise so we can ensure that each student got the latest lab files.

Having a different branch for each revision may also lead to just as many issues.

**Comment by vitaly-livshits** [comment]
@seesharprun I've opened the ticket for LODS last night. It's a bandaid solution.

I've also told my students this week to replace checkout of master with getting the v20190725.03 label. It's very awkward to have to catch this type of issue during live delivery. Surely there is a better way.

IMO for a stable solution Microsoft needs to take ownership of the process. It needs to deliver very specific and reliable instructions to ALHs so ALHs can host a quality lab experience. For a start, the lab instructions **must** match the lab files. It's an entirely solvable change management scenario.

There are so many options. Either package the instructions with the files, similar to the way "legacy" MOCs had lab files at https://www.microsoft.com/en-us/learning/companion-moc.aspx that matched the instructions in Skillpipe. Or pull everything from GitHub in realtime during class like @tiagocostapt says (although even that's susceptible to intra-lab changes). Or have concrete releases here on GitHub like I was originally suggesting.

The current mix-and-match approach is simply too fragile and it completely ignores breaking changes and caching of instructions by ALHs. Keep in mind, even if ALH pulls instructions from GitHub the morning of the delivery, the files could be updated in the master branch here a few minutes thereafter. And then the entire lab is broken again.

This aspect really needs to be thought through a bit more by MSL. Otherwise it's very challenging to have a great course experience. @seesharprun, as the repo maintainer I am hopeful you could lead this discussion with MSL. Thanks!

**Comment by seesharprun** [comment]
We do have concrete releases of Allfiles content and lab instructions on GitHub. They can be found here <https://github.com/MicrosoftLearning/AZ-203-DevelopingSolutionsforMicrosoftAzure/releases> and it's the first link on the homepage for the repository.

Today was the first time I learned how the ALHs get the lab instructions. The process is definitely flawed. Unfortunately, I do not have a contact at Microsoft to escalate this issue.

As a maintainer for the repository, I do need to solve this. One way I could do it would be to remove the instructions to clone the Allfiles content. This would force MCTs/ALHs to use the Allfiles content and lab instructions available on GitHub. That's the direction I lean on this issue right now.

@vitaly-livshits & @tiagocostapt, would removing the cloning instructions work for you?

**Comment by vitaly-livshits** [comment]
@seesharprun , I would be hesitant to remove the cloning instructions to solve the change management issues. Why? At least two reasons:

1) It would add another steps for ALHs to manually get the files from the repo. Even if they know and remember to get the files, they would need to do so at the right point in time, concurrently with instructions. Fragility due to synchronization issues would come back. It could end up being even more fragile than the current workflow.

2) We would lose the ability to (easily) patch files live on GitHub in case of any urgent issues in the labs. The files will be static on the ALH VMs during delivery and thus updates will need to be done by students, likely with MCT guidance. Another manual step in the process and it would offload the responsibility for functioning labs to the customers (MCTs & students) instead of providers (Microsoft & ALHs).

How about a slight change to instructions, to make them clone files from a tag (or another stable versioning mechanism) instead of getting the latest from the master? The tag would then be set to the corresponding set of instructions that are guaranteed to work with the files. 

This could probably be automated within the GitHub build system. The idea is to have GitHub produce a consistent bundle ("build") that works as a whole. The ALHs will then take this build and it would just work. I think it will stay functioning in all circumstances and if designed properly it would even allow patches to be released automatically into the labs. Do you see any issues with this approach?

Bottom line: any solution must give MCTs and students labs that work as written. Zero tweaks should be necessary during delivery to assure lab quality. Both MCTs and students have enough on their plates as it is during the course.

_P.S. I am still very puzzled how this is all being discussed and designed by MCTs like you and me instead of by MSL. MSL owns this MOC course. MSL would thus expected to deliver ready-to-use labs that "just work". I suppose we could leave that conversation for another day :)_

**Comment by tiagocostapt** [comment]
@seesharprun my advise is to not remove the git clone instructions. If ALHs don't have the labs updated and this generate conflicts with the allfiles they should remove the git clone instructions an pre-download the allfiles folder into their system. 
Having the git clone instructions is very useful to keep students with the latest version of the allfiles. Sometimes the repo is updated in the middle of a delivery and this way students get the most updated version when they start a specific lab.

**Comment by seesharprun** [comment]
I have had a couple of discussions with testers and we have a big hesitation to adding version tags and strings to the clone instructions:

- If any MCT decides to create a PR, there's no way for them to tell what version number to update the git clone instructions to
- If someone forgets to update the clone instructions, it breaks the lab

Again, the clone instructions only exist because I can't force the ALH to download the Allfiles content in the release section. The Allfiles content has been in the release section since the launch of the course.

**Comment by MikeBazMSFT** [comment]
Some hosters - including LODS - have the ability to automatically update their instructions from GitHub, although some "niceness" is lost (e. g. in the LODS case, some of the lab step tracking stuff is lost) - I'd rather have up-to-date instructions than some checkboxes, myself, since they don't have the checkboxes going to GitHub directly anyway.

For AllFIles, some providers can run commands as part of starting a VM (95% sure LODS can do this for example), so the VM can just do the clone on its own - or you could have it as an automatic task on login to run it so that students don't have to mess with it.  While I agree with the idea that "it's cool to show them GitHub," there's so much going on in this class that anything we can do to avoid cognitive overload is good.

**Comment by seesharprun** [comment]
I'm leaning heavily towards just following the model of the other courses on GitHub and letting the ALHs worry about keeping the AllFiles content up to date.

Anyone who wants can still easily figure out how to clone the AllFiles if they want up-to-the-minute lab files.
